src/generators.rs

```diff
@@ -985,6 +985,101 @@ pub fn directed_grid_graph(
     })
 }
 
+/// Generate an undirected hexagonal graph.
+///
+/// :param int rows: The number of rows to generate the graph with.
+/// :param list cols: The number of columns to generate the graph with.
+/// :param bool multigraph: When set to False the output
+///     :class:`~retworkx.PyGraph` object will not be not be a multigraph and
+///     won't  allow parallel edges to be added. Instead
+///     calls which would create a parallel edge will update the existing edge.
+///
+/// :returns: The generated hexagonal graph
+/// :rtype: PyGraph
+/// :raises IndexError: If neither ``rows`` or ``cols`` are
+///      specified
+///
+/// .. jupyter-execute::
+///
+///   import os
+///   import tempfile
+///
+///   import pydot
+///   from PIL import Image
+///
+///   import retworkx.generators
+///
+///   graph = retworkx.generators.hexagonal_graph(2, 2)
+///   dot_str = graph.to_dot(
+///       lambda node: dict(
+///           color='black', fillcolor='lightblue', style='filled'))
+///   dot = pydot.graph_from_dot_data(dot_str)[0]
+///
+///   with tempfile.TemporaryDirectory() as tmpdirname:
+///       tmp_path = os.path.join(tmpdirname, 'dag.png')
+///       dot.write_png(tmp_path)
+///       image = Image.open(tmp_path)
+///       os.remove(tmp_path)
+///   image
+///
+#[pyfunction(multigraph = true)]
+#[text_signature = "(/, rows=None, cols=None, multigraph=True)"]
+pub fn hexagonal_graph(
+    py: Python,
+    rows: usize,
+    cols: usize,
+    multigraph: bool,
+) -> PyResult<graph::PyGraph> {
+    let mut graph = StableUnGraph::<PyObject, PyObject>::default();
+
+    let mut rowlen = rows;
+    let mut collen = cols;
+
+    // Needs two times the number of nodes vertically
+    rowlen = 2 * rowlen + 2;
+    collen += 1;
+    let num_nodes = rowlen * collen;
+
+    let nodes: Vec<NodeIndex> =
+        (0..num_nodes).map(|_| graph.add_node(py.None())).collect();
+
+    // Add column edges
+    for i in 0..collen {
+        for j in 0..(rowlen - 1) {
+            graph.add_edge(
+                nodes[i * rowlen + j],
+                nodes[i * rowlen + (j + 1)],
+                py.None(),
+            );
+        }
+    }
+
+    // Add row edges
+    for i in 0..(collen - 1) {
+        for j in 0..rowlen {
+            if i % 2 == j % 2 {
+                graph.add_edge(
+                    nodes[i * rowlen + j],
+                    nodes[(i + 1) * rowlen + j],
+                    py.None(),
+                );
+            }
+        }
+    }
+
+    // Remove corner nodes
+    graph.remove_node(nodes[rowlen - 1]);
+    graph.remove_node(
+        nodes[(collen - 1) * rowlen + (rowlen - 1) * ((collen - 1) % 2)],
+    );
```

Comment:
Instead of removing the corner nodes here which will result in a graph with holes in the node indices couldn't we just detect the corner nodes in the edge loop and not add the edge?

Reply:
Sure. I will do an implementation without removing nodes.

Reply:
Actually deleting the nodes is required for the implementation. Thus, we have two options:

- Documenting that we are deleting nodes and which nodes will be missing;
- Reindexing all nodes.

I will take a look at what will be the overhead of reindexing all nodes.

Reply:
I ran a benchmark comparing reindexing nodes:

| Branches        | Time           | 
| ------------- |:-------------:| 
| [without reindexing](https://github.com/nahumsa/retworkx/tree/Expand_Generators)      | 0.115 |
| [with reindexing](https://github.com/nahumsa/retworkx/tree/hex-reindexing)     |   1.043  | 

Reindexing this graph adds a significant overhead, therefore I will choose to document which nodes that we are deleting.